src/sage/schemes/elliptic_curves/ell_finite_field.py

```diff
@@ -1122,14 +1122,53 @@ def torsion_basis(self, n):
         :meth:`AdditiveAbelianGroupWrapper.torsion_subgroup`.
         """
         # TODO: In many cases this is not the fastest algorithm.
-        # Alternatives include factoring division polynomials and
+        # Alternatives include factoring division polynomials (now implemented) and
         # random sampling (like PARI's ellgroup, but with a milder
-        # termination condition). We should implement these too
+        # termination condition). We should implement the latter too
         # and figure out when to use which.
-        T = self.abelian_group().torsion_subgroup(n)
-        if T.invariants() != (n, n):
-            raise ValueError(f'curve does not have full rational {n}-torsion')
-        return tuple(P.element() for P in T.gens())
+        if algorithm == "abelian_group":
+            T = self.abelian_group().torsion_subgroup(n)
+            if T.invariants() != (n, n):
+                raise ValueError(f'curve does not have full rational {n}-torsion')
+            return tuple(P.element() for P in T.gens())
+        
+        elif algorithm == "division_polynomial":
```

Comment:
Currently if a user supplies a wrong algorithm argument this code will fail without proper error handling. See https://github.com/sagemath/sage/blob/3dd953c3aa6b5143071b6c39208199cf128c8080/src/sage/schemes/elliptic_curves/ell_point.py#L1821 for inspiration on how to handle the user input.